src/annotator/guest.js

```diff
@@ -315,6 +327,39 @@ export default class Guest {
     });
   }
 
+  async _connectHostEvents() {
+    this._hostRPC.on(
+      'unselectTextExceptIn',
+      /** @param {string} frameIdentifier */
+      frameIdentifier => {
+        if (this._frameIdentifier === frameIdentifier) {
+          return;
+        }
+
+        this._onClearSelection({ informHostFrame: false });
+        // It would be nice to clear the shadow of the selection calling
```

Comment:
I'm not sure what is meant by "shadow of the selection"? Do you mean the way some browsers render the selection in un-focused iframes in a grey color? That's not a shadow of the selection, that _is_ the selection.

A way to avoid this could be to set a flag to ignore the next selection update, which is then reset when the update is received. This would avoid the need to call `_onClearSelection` here explicitly.

Reply:
A sketch of how this might work:

```diff
diff --git a/src/annotator/guest.js b/src/annotator/guest.js
index c4eb66b0f..ea0905894 100644
--- a/src/annotator/guest.js
+++ b/src/annotator/guest.js
@@ -90,6 +90,10 @@ function resolveAnchor(anchor) {
   }
 }
 
+function clearSelection() {
+  /** @type {Selection} */ (document.getSelection()).removeAllRanges();
+}
+
 /**
  * `Guest` is the central class of the annotator that handles anchoring (locating)
  * annotations in the document when they are fetched by the sidebar, rendering
@@ -136,11 +140,11 @@ export default class Guest {
     this._adder = new Adder(this.element, {
       onAnnotate: async () => {
         await this.createAnnotation();
-        /** @type {Selection} */ (document.getSelection()).removeAllRanges();
+        clearSelection();
       },
       onHighlight: async () => {
         await this.createAnnotation({ highlight: true });
-        /** @type {Selection} */ (document.getSelection()).removeAllRanges();
+        clearSelection();
       },
       onShowAnnotations: anns => {
         this.selectAnnotations(anns);
@@ -154,6 +158,7 @@ export default class Guest {
         this._onClearSelection();
       }
     });
+    this._ignoreNextSelection = false;
 
     /**
      * The anchors generated by resolving annotation selectors to locations in the
@@ -335,12 +340,8 @@ export default class Guest {
         if (this._frameIdentifier === frameIdentifier) {
           return;
         }
-
-        this._onClearSelection({ informHostFrame: false });
-        // It would be nice to clear the shadow of the selection calling
-        // `document.getSelection()?.removeAllRanges();` however, because of the
-        // selection observer, that will trigger a call to `_onClearSelection({informHostFrame: true})`
-        // which in turn will remove the selection from the `Guest` that has the selection.
+        this._ignoreNextSelection = true;
+        clearSelection();
       }
     );
 
@@ -661,12 +662,15 @@ export default class Guest {
     this._adder.show(focusRect, isBackwards);
   }
 
-  _onClearSelection({ informHostFrame = true } = {}) {
+  _onClearSelection() {
     this._isAdderVisible = false;
     this._adder.hide();
     this.selectedRanges = [];
-    if (informHostFrame) {
+
+    if (!this._ignoreNextSelection) {
       this._hostRPC.call('textUnselectedIn', this._frameIdentifier);
+    } else {
+      this._ignoreNextSelection = false;
     }
   }
 ```

Reply:
I believe this proposal won't resolve the problem.

I propose we leave this as a follow up.

Reply:
> I believe this proposal won't resolve the problem.

Why not? I did test this code locally.